### PR TITLE
[Polygon Drawing] Render polygons on map

### PR DESCRIPTION
**Pull request by HemanParbhakar**
Contributes #358 

Polygon loaded from local db and shown on Map.

@gino-m @shobhitagarwal1612  PTAL?

### CODE REVIEW

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 1]
```diff
@@ -198,6 +205,17 @@ private void handleAmbiguity(LatLng latLng) {
       }
     }
 
+    for (Map.Entry<MapFeature, List<LatLng>> json : polygons.entrySet()) {
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:208]: I was thrown off by the variable name `json` here, thought I hadn't had enough coffee yet :) Items in `entrySet` aren't json, they're map entries. Could you please rename to `entry` here? 

Also, to avoid confusion, please assign the key and value to local variables `mapFeature` and `vertices`.

Would you mind also updating the above code block in a separate commit as well?

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 2]
```diff
@@ -277,28 +295,45 @@ private void addMapPin(MapPin mapPin) {
...
-    for (ImmutableSet<Point> vertices : mapPolygon.getVertices()) {
-      PolylineOptions options = new PolylineOptions();
-
-      // Read-only
-      options.clickable(false);
-
-      // Add vertices to PolylineOptions
-      stream(vertices).map(GoogleMapsMapAdapter::toLatLng).forEach(options::add);
+    PolylineOptions options = new PolylineOptions();
+    // Read-only
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:299]: Please remove this comment, since it doesn't add/clarify anything.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 3]
```diff
@@ -277,28 +295,45 @@ private void addMapPin(MapPin mapPin) {
...
+  }
 
-      // Style polyline
-      polyline.setStartCap(new RoundCap());
-      polyline.setEndCap(new RoundCap());
-      polyline.setWidth(getPolylineStrokeWidth());
-      polyline.setColor(parseColor(mapPolygon.getStyle().getColor()));
-      polyline.setJointType(JointType.ROUND);
+  private boolean isPolygonCompleted(List<Point> vertices) {
+    return vertices.size() != stream(vertices).distinct().count();
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:324]: Not sure this is correct; "In geometry, a polygon (/ˈpɒlɪɡɒn/) is a plane figure that is described by a finite number of straight line segments connected to form a closed polygonal chain (or polygonal circuit). "
https://en.wikipedia.org/wiki/Polygon

In that case, we should be looking if the first and last vertices are equivalent, not if all vertices are distinct as you're doing here. Also it implies the method should be called "isClosedPolygonalChain", with a comment that clarifies that it's checking if "the first and last vertices match such that they form a closed polygonal chain."

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 4]
```diff
@@ -277,28 +295,45 @@ private void addMapPin(MapPin mapPin) {
...
-
-      // Read-only
-      options.clickable(false);
-
-      // Add vertices to PolylineOptions
-      stream(vertices).map(GoogleMapsMapAdapter::toLatLng).forEach(options::add);
+    PolylineOptions options = new PolylineOptions();
+    // Read-only
+    options.clickable(false);
+    // Add vertices to PolylineOptions
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:301]: This comment is in the wrong place; in this step we're converting the vertices to a list of LatLng (no comment needed unless it's clarifying what's already clear in code).

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 5]
```diff
@@ -277,28 +295,45 @@ private void addMapPin(MapPin mapPin) {
...
+    ImmutableList<LatLng> vertices = stream(mapPolygon.getVertices())
+        .map(GoogleMapsMapAdapter::toLatLng)
+        .collect(toImmutableList());
+    stream(vertices).forEach(options::add);
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:305]: Please use `options.addAll(vertices)` instead

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 6]
```diff
@@ -277,28 +295,45 @@ private void addMapPin(MapPin mapPin) {
...
+    // Add to map
+    Polyline polyline = map.addPolyline(options);
+    polyline.setTag(mapPolygon);
+    // Style polyline
+    if (!isPolygonCompleted(mapPolygon.getVertices())) {
+      BitmapDescriptor customCap = bitmapDescriptorFromVector();
+      polyline.setStartCap(new CustomCap(customCap));
+      polyline.setEndCap(new CustomCap(customCap));
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:313]: Do we need to contsruct two instances of CustomCap for each polygon? Or would one for the whole GoogleMapsMapAdapter suffice?

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 7]
```diff
@@ -277,28 +295,45 @@ private void addMapPin(MapPin mapPin) {
...
+  private boolean isPolygonCompleted(List<Point> vertices) {
+    return vertices.size() != stream(vertices).distinct().count();
+  }
 
-      polylines.add(polyline);
-    }
+  private BitmapDescriptor bitmapDescriptorFromVector() {
+    Drawable vectorDrawable = ContextCompat.getDrawable(context, R.drawable.ic_endpoint);
+    vectorDrawable
+        .setBounds(4, 4, vectorDrawable.getIntrinsicWidth(), vectorDrawable.getIntrinsicHeight());
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:330]: Why `(4, 4)` specifically? It's not clear looking at the code; please either comment or move to dimens.

**Review by gino-m — CHANGES_REQUESTED** [review]

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java (hunk 6 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:313]: yes would work so shall we create this on constructor?

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java (hunk 7 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java]: updated

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java (hunk 5 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java]: updated

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java (hunk 4 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java]: updated.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java (hunk 3 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java]: updated.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java (hunk 2 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java]: updated.

File: gnd/src/main/java/com/google/android/gnd/ui/home/mapcontainer/MapContainerViewModel.java [hunk 8]
```diff
@@ -199,16 +201,27 @@ private void onProjectChange(Optional<Project> project) {
             .map(MapContainerViewModel::toMapPin)
             .collect(toImmutableSet());
 
-    // TODO: Add support for polylines and polygons similar to mapPins
+    // TODO: Add support for polylines similar to mapPins
```
[the remark this answers is not in the document]
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/home/mapcontainer/MapContainerViewModel.java]: updated.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 9]
```diff
@@ -296,13 +296,12 @@ private void addMapPin(MapPin mapPin) {
...
   private void addMapPolyline(MapPolygon mapPolygon) {
     PolylineOptions options = new PolylineOptions();
-    // Read-only
     options.clickable(false);
-    // Add vertices to PolylineOptions
     ImmutableList<LatLng> vertices = stream(mapPolygon.getVertices())
         .map(GoogleMapsMapAdapter::toLatLng)
         .collect(toImmutableList());
-    stream(vertices).forEach(options::add);
+    // Add vertices to PolylineOptions
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:303]: Redundant comment, please remove.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 10]
```diff
@@ -321,13 +320,17 @@ private void addMapPolyline(MapPolygon mapPolygon) {
   }
 
   private boolean isPolygonCompleted(List<Point> vertices) {
-    return vertices.size() != stream(vertices).distinct().count();
+    return vertices.get(vertices.size() - 1) == vertices.get(0);
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:323]: This will force quit if `vertices` is empty.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 11]
```diff
@@ -321,13 +320,17 @@ private void addMapPolyline(MapPolygon mapPolygon) {
...
   }
 
   private BitmapDescriptor bitmapDescriptorFromVector() {
     Drawable vectorDrawable = ContextCompat.getDrawable(context, R.drawable.ic_endpoint);
+    // Specify a bounding rectangle for the Drawable.
     vectorDrawable
-        .setBounds(4, 4, vectorDrawable.getIntrinsicWidth(), vectorDrawable.getIntrinsicHeight());
+        .setBounds((int) context.getResources().getDimension(R.dimen.polyline_endpoint_bound),
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:330]: Can we just trim these pixels off the vector instead of clipping them here?

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 12]
```diff
@@ -277,28 +295,48 @@ private void addMapPin(MapPin mapPin) {
...
+
+    polylines.add(polyline);
+    polygons.put(mapPolygon, vertices);
+  }
+
+  private boolean isPolygonCompleted(List<Point> vertices) {
+    return vertices.get(vertices.size() - 1) == vertices.get(0);
+  }
+
+  private BitmapDescriptor bitmapDescriptorFromVector() {
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:326]: See comment in previous review about refactoring part of this into BitmapUtil.

**Review by gino-m — CHANGES_REQUESTED** [review]

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java (hunk 11 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java]: reduced the size of svg to 16px so now no clipping required.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java (hunk 10 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:325]: added check if vertices size is less than 3 return false

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java (hunk 9 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java]: removed.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java (hunk 12 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java]: moved to BitmapUtil.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java (hunk 6 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java]: added in constructor.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java (hunk 1 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java]: renamed.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 13]
```diff
@@ -137,13 +145,17 @@
...
   private int cameraChangeReason = REASON_DEVELOPER_ANIMATION;
 
-  public GoogleMapsMapAdapter(GoogleMap map, Context context, MarkerIconFactory markerIconFactory) {
+  public GoogleMapsMapAdapter(GoogleMap map, Context context,
+      MarkerIconFactory markerIconFactory,
+      BitmapUtil bitmapUtil) {
     this.map = map;
     this.context = context;
     this.markerIconFactory = markerIconFactory;
+    this.customCap = new CustomCap(bitmapUtil.bitmapDescriptorFromVector());
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:158]: From which vector? I wouldn't expect BitmapUtil to have which vector to convert hard-coded. Perhaps you want to pass it as an arg here?

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 14]
```diff
@@ -277,28 +302,33 @@ private void addMapPin(MapPin mapPin) {
...
-      // Add to map
-      Polyline polyline = map.addPolyline(options);
-      polyline.setTag(mapPolygon);
+    PolylineOptions options = new PolylineOptions();
+    options.clickable(false);
+    ImmutableList<LatLng> vertices = stream(mapPolygon.getVertices())
+        .map(GoogleMapsMapAdapter::toLatLng)
+        .collect(toImmutableList());
+    options.addAll(vertices);
+    // Add to map
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:311]: "Add to map" just repeats what's already on the next line ("map.addPolyline"); please only use comments when it's not possible to make your code self-explanatory. From Robert C. Martin's Clean Code:

> The proper use of comments is to compensate for our failure to express ourself in code. Note that I used the word failure. I meant it. Comments are always failures. We must have them because we cannot always figure out how to express ourselves without them, but their use is not a cause for celebration.... Every time you express yourself in code, you should pat yourself on the back. Every time you write a comment, you should grimace and feel the failure of your ability of expression.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java [hunk 15]
```diff
@@ -277,28 +302,33 @@ private void addMapPin(MapPin mapPin) {
...
+    Polyline polyline = map.addPolyline(options);
+    polyline.setTag(mapPolygon);
+    // Style polyline
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:314]: Same here.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java (hunk 10 above)
Comment (gino-m) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapAdapter.java:331]: -> `return vertices.size() > 2 && vertices.get(vertices.size() - 1) == vertices.get(0)`

Also, as always please be sure to run `google-java-format` on code before submitting.

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapProvider.java [hunk 16]
```diff
@@ -60,7 +63,9 @@ private void createMapAsync() {
         .getMapAsync(
             googleMap ->
                 map.onSuccess(
-                    new GoogleMapsMapAdapter(googleMap, fragment.getContext(), markerIconFactory)));
+                    new GoogleMapsMapAdapter(googleMap,
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapProvider.java:66]: google-java-format please.

File: gnd/src/main/java/com/google/android/gnd/ui/util/BitmapUtil.java [hunk 17]
```diff
@@ -33,8 +39,26 @@
...
-  /** Retrieves an image for the given url as a {@link Bitmap}. */
+  /**
+   * Retrieves an image for the given url as a {@link Bitmap}.
+   */
   public Bitmap fromUri(Uri url) throws IOException {
     return Media.getBitmap(context.getContentResolver(), url);
   }
+
+  public BitmapDescriptor bitmapDescriptorFromVector() {
+    Drawable vectorDrawable = ContextCompat.getDrawable(context, R.drawable.ic_endpoint);
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/ui/util/BitmapUtil.java:50]: BitmapUtil is a generic utility, but `ic_endpoint` is specific to polygon drawing and thus doesn't belong hardcoded here. Please pass as an arg instead.

**Review by gino-m — CHANGES_REQUESTED** [review]

File: gnd/src/main/java/com/google/android/gnd/ui/util/BitmapUtil.java (hunk 17 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/util/BitmapUtil.java]: updated

File: gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapProvider.java (hunk 16 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/ui/map/gms/GoogleMapsMapProvider.java]: updated

File: gnd/build.gradle [hunk 18]
```diff
@@ -152,14 +152,27 @@ android {
...
+                option("NullAway:ExcludedClasses",
+                    "com.google.android.gnd.ui.observationdetails.ObservationDetailsFragmentArgs," +
+                        "com.google.android.gnd.ui.observationdetails.ObservationDetailsFragmentDirections," +
+                        "com.google.android.gnd.ui.editobservation.EditObservationFragmentArgs," +
+                        "com.google.android.gnd.ui.home.HomeScreenFragmentArgs," +
+                        "com.google.android.gnd.ui.home.HomeScreenFragmentDirections," +
+                        "com.google.android.gnd.ui.offlinebasemap.viewer.OfflineBaseMapsFragmentArgs," +
+                        "com.google.android.gnd.ui.offlinebasemap.OfflineBaseMapsFragmentDirections," +
+                        "com.google.android.gnd.ui.offlinebasemap.viewer.OfflineBaseMapViewerFragmentArgs"
+                )
```
Comment (shobhitagarwal1612) [inline gnd/build.gradle:175]: Please revert these changes. It probably happened during merge conflicts.

Comment (HemanParbhakar) [reply gnd/build.gradle]: reverted

**Review by gino-m — APPROVED** [review]
LGTM!
